### ISSUE TITLE
Fix false negatives for multiline selectors in indentation

### ISSUE THREAD

**Issue by ThomasGreiner** [stylelint/stylelint#3461]
> What is the problem you're trying to solve?

Some of the selectors we use in one of our projects are quite long and thereby violate the `max-line-length: 80` rule.

```css
body:not([data-dialog="language-add"]):not([data-dialog="language-change"]) #dialog-content-language-add
```

We tend to resolve such issues by splitting up the line and indenting subsequent lines.

```css
body:not([data-dialog="language-add"]):not([data-dialog="language-change"])
  #dialog-content-language-add
```

Because the alternative would be to not indent subsequent lines which, however, makes it diffcult to differentiate whether a line is a selector or merely a continuation of the previous selector.

```
body:not([data-dialog="language-add"]):not([data-dialog="language-change"])
#dialog-content-language-add
```

However, this results in a violation of our `indentation: 2` rule and there doesn't appear to be an option for excluding selectors 

```
✖  Expected indentation of 0 spaces   indentation
```

Because of that and https://github.com/stylelint/stylelint/issues/3111 we currently have to disable the "indentation" rule for the entire block which is not ideal.

```css
/* stylelint-disable indentation */
body:not([data-dialog="language-add"]):not([data-dialog="language-change"])
  #dialog-content-language-add
{
  display: none;
}
/* stylelint-enable indentation */
```

Furthermore, due to https://github.com/stylelint/stylelint/issues/2643, that means that we have to be extra careful when running `stylelint --fix`.

> What solution would you like to see?

A new option for the "indentation" rule that allows lines that represent continuations of selectors to be indendent by one more level.

**Comment by jeddy3** [comment]
@ThomasGreiner Thanks for the request and for using the appropriate template.

Yes, I can see how indenting in these cases can make the code my readable.

To help us name such an option correctly, can you tell us what the coding style would be if it was a selector list. For example, would it be:

```css
body:not([data-dialog="language-add"]):not([data-dialog="language-change"])
  #dialog-content-language-add,
body:not([data-dialog="position-left"]):not([data-dialog="position-left"])
  #dialog-content-position-add
{
  display: none;
}
```

Or something else?

**Comment by ThomasGreiner** [comment]
> To help us name such an option correctly, can you tell us what the coding style would be if it was a selector list. For example, would it be:

@jeddy3 Yes, that style would be ideal if that's possible.

**Comment by jeddy3** [comment]
Groovy. In that case, it feels similar to the existing [`indentClosingBrace` option](https://github.com/stylelint/stylelint/tree/master/lib/rules/indentation#indentclosingbrace-truefalse), rather than the [`"except: []` one](https://github.com/stylelint/stylelint/tree/master/lib/rules/indentation#except-block-param-value).

I think an `indentMultiLineSelector: true| false` (default is `false`) option will work here. (Everyone feel free jump in with objections, though).

I'll label this up as "help wanted" and open this out to a PR. There's [a section in the Developer Guide](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#adding-an-option-to-an-existing-rule) on how to get started.

**Comment by WebReflection** [comment]
FYI I've created a function to monkey patch spaces **after** passing through `--fix`

```js
function reSpace(code, indent) {
  var spaces = new Array((indent || 2) + 1).join(' ');
  var comments = [];
  // temporarily drop all multiline comments
  var lines = code.replace(
    /\/\*[\s\S]*?\*\//g,
    function ($0) {
      // surround indexes with the lovely SOH char
      return '\x01' + (comments.push($0) - 1) + '\x01';
    }
  ).split('\n');
  for (var line, prev, j, i = lines.length; i--;) {
    line = lines[i].replace(/^\s+|\s+/g, '');
    if (/\{$/.test(line)) {
      while (i > 0) {
        j = line.length === 1 ? 2 : 1;
        if ((i - j) >= 0) {
          prev = lines[i - j--].replace(/^\s+|\s+/g, '');
          if (
            prev.length &&
            !/[,\x01]$/.test(prev) &&
            !/^\/\//.test(prev)
          ) {
            lines[i - j] = spaces + lines[i - j];
          }
        }
        i--;
      }
    }
  }
  // put multiline comments back
  return lines.join('\n').replace(
    /\x01(\d+)\x01/g,
    function ($0, $1) {
      return comments[$1];
    }
  );
}
```

This function should put back those cases where the selector is at the beginning of the line but there is no comma before.

Test case:

```js
// test
console.assert(
  reSpace(
`/* this is a test */
#some-very-long-selector
.not-completed-in-previous-line,
#some-small-selector,
.some-other-long-selector
[also="not-completed"],
// some random comment
#and-finally
.another-long-selector
{
  color: green;
}`
  ) ===
`/* this is a test */
#some-very-long-selector
  .not-completed-in-previous-line,
#some-small-selector,
.some-other-long-selector
  [also="not-completed"],
// some random comment
#and-finally
  .another-long-selector
{
  color: green;
}`
);


console.assert(
  reSpace(
`/* this is a test */
#some-very-long-selector
.not-completed-in-previous-line,
#some-small-selector,
.some-other-long-selector
[also="not-completed"],
// some random comment
#and-finally
.another-long-selector {
  color: green;
}`
  ) ===
`/* this is a test */
#some-very-long-selector
  .not-completed-in-previous-line,
#some-small-selector,
.some-other-long-selector
  [also="not-completed"],
// some random comment
#and-finally
  .another-long-selector {
  color: green;
}`
);
```

I am sure this is not the best / long term solution but for those who need a quick fix I have hooked the invoke of such function right after `stylelint --fix $file`

Hope this helps.

**Comment by jeddy3** [comment]
@WebReflection Thanks for sharing.

> I am sure this is not the best / long term solution

I believe adding a `indentMultiLineSelector` option to the rule would do it.

**Comment by fregante** [comment]
Would this issue also apply indentation to this selector?

``` css
:is(
a,
b
) {}
``` 

I’d expect lines inside the parens to be indented

``` css
:is(
  a,
  b
) {}
```

**Comment by jeddy3** [comment]
@fregante I believe that's a separate problem. Can you create a new bug issue for it, please?

I think the following should be the default behaviour of the rule:

```css
:is(
  a,
  b
) {}
```

We [apply indentation to parens in at-rules and declaration-values](https://stylelint.io/user-guide/rules/indentation#indentinsideparens-twiceonce-at-root-twice-in-block), but not those in selectors. It should, though.

---

Revisiting the original issue, I think indenting multiline selectors should be the default behaviour rather than an option. I'll change the labelling accordingly.

When both issues are fixed, we end up with:

```css
body:not([data-dialog="language-add"]) .class
  :not([data-dialog="language-change"])
  :is(
    a,
    b
  )
  #dialog-content-language-add {}
```

**Comment by jeddy3** [comment]
Closing [as stylistic rules are frozen](https://github.com/stylelint/stylelint/issues/5674). The community is welcome to migrate the rule [to a plugin](https://stylelint.io/developer-guide/plugins) and fix this bug.
